From ac865c51137dad2692fa1a8a7cee1a6d575f5c39 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Tue, 19 Aug 2025 19:55:05 +0000
Subject: [PATCH 1/3] [http2] set error on zero return from write

Sometimes servers try to close (close_notify alert) the TLS connection
before all data was transmitted. This results in 0 returned for every
write. That indicates TLS connection is closed(SSL_write returns SSL_ERROR_ZERO_RETURN),
but underlying socket connection is still open. Resulting in infinite loop in io_try_write.
---
 libsoup/http2/soup-client-message-io-http2.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/libsoup/http2/soup-client-message-io-http2.c b/libsoup/http2/soup-client-message-io-http2.c
index 1e8e8a71..79027ecd 100644
--- a/libsoup/http2/soup-client-message-io-http2.c
+++ b/libsoup/http2/soup-client-message-io-http2.c
@@ -372,6 +372,13 @@ io_write (SoupClientMessageIOHTTP2 *io,
         if (ret < 0)
                 return FALSE;
 
+        if (ret == 0) {
+                g_set_error_literal (error, G_IO_ERROR,
+                                     G_IO_ERROR_PARTIAL_INPUT,
+                                     _("Connection terminated unexpectedly"));
+                return FALSE;
+        }
+
         io->written_bytes += ret;
         return TRUE;
 }
-- 
2.34.1

